handlers/mtn/mtn.go

```diff
@@ -0,0 +1,245 @@
+package mtn
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/gomodule/redigo/redis"
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/pkg/errors"
+)
+
+var (
+	sendURL      = "https://api.mtn.com/v2/messages/sms/outbound"
+	maxMsgLength = 160
+	tokenURL     = "https://api.mtn.com/v1/oauth/access_token?grant_type=client_credentials"
+)
+
+func init() {
+	courier.RegisterHandler(newHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+
+	fetchTokenMutex sync.Mutex
+}
+
+func newHandler() courier.ChannelHandler {
+	return &handler{
+		BaseHandler:     handlers.NewBaseHandler(courier.ChannelType("MTN"), "MTN Developer Portal"),
+		fetchTokenMutex: sync.Mutex{},
+	}
+}
+
+// Initialize implements courier.ChannelHandler
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	s.AddHandlerRoute(h, http.MethodPost, "receive", h.receiveMessage)
+	s.AddHandlerRoute(h, http.MethodPost, "status", h.receiveStatus)
+	return nil
+}
+
+var statusMapping = map[string]courier.MsgStatusValue{
+	"DeliveredToTerminal": courier.MsgDelivered,
+	"DeliveryUncertain":   courier.MsgSent,
+	"DeliveryImpossible":  courier.MsgErrored,
+	"DeliveredToNetwork":  courier.MsgSent,
+
+	// no changes
+	"MessageWaiting":                   courier.MsgWired,
+	"DeliveryNotificationNotSupported": courier.MsgWired,
+}
+
+type statusPayload struct {
+	RequestID      string `json:"requestId"`
+	DeliveryStatus []struct {
+		ReceiverAddress string `json:"receiverAddress"`
+		Status          string `json:"status"`
+	} `json:"deliveryStatus"`
+}
+
+// receiveStatus is our HTTP handler function for status updates
+func (h *handler) receiveStatus(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request, clog *courier.ChannelLog) ([]courier.Event, error) {
+	payload := &statusPayload{}
+	err := handlers.DecodeAndValidateJSON(payload, r)
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+	msgStatus, found := statusMapping[payload.DeliveryStatus[0].Status]
+	if !found {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r,
+			fmt.Errorf("unknown status '%s', must be one of 'DeliveredToTerminal','DeliveryUncertain','DeliveryImpossible','DeliveredToNetwork', 'MessageWaiting', or 'DeliveryNotificationNotSupported'", payload.DeliveryStatus[0].Status))
```

Comment:
I don't think you need to list all statuses here.. this error message is only for us to debug and we know what the accepted statuses are